sympy/core/add.py

```diff
@@ -870,10 +870,29 @@ def _eval_as_leading_term(self, x):
             return expr.as_leading_term(x)
 
         infinite = [t for t in expr.args if t.is_infinite]
+        compute = False
 
-        expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
-        if not expr:
-            # simple leading term analysis gave us 0 but we have to send
+        leading_term = [t.as_leading_term(x) for t in expr.args]
+        new_expr = expr.func(*leading_term).removeO()
+
+        if not new_expr:
+            compute = True
+        elif new_expr is not S.NaN:
+            if new_expr.is_Add:
+                final_leading_term = [t for t in new_expr.args]
+            else:
+                final_leading_term = [new_expr]
+
+            if len(expr.args) != len(final_leading_term):
+                canceled_term = [t for t in leading_term if t not in final_leading_term]
+                expr_sum = expr.func(*canceled_term)
```

Comment:
What is the purpose of this? The `canceled_term` list could contain all leading terms on the original list and then nothing would have been gained.

Reply:
This will only enter when **we have atleast one term** in new_expr. 
So, i don't think `cancelled_term list` could contain all leading terms on the original list.
Because I have put that condition` if compute == False:`